### PR TITLE
[receiver/kafkametrics] add support for topic configurations and cluster_alias field

### PR DESCRIPTION
**Pull request by rshiyani**
**Description:** <Describe what has changed.>
1. Adding support for kafka topic and broker configurations. receiver will expose following metrics.
- `kafka.broker.log_retention_period`
- `kafka.topic.replication_factor`
- `kafka.topic.log_retention_period`
- `kafka.topic.log_retention_size`
- `kafka.topic.min_insync_replicas`

2. Add an optional configuration `cluster_alias` field. 
3. assign a new metric attribute `cluster_alias` to all metrics.

**Link to tracking Issue:** <Issue number if applicable>
#34148 

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: receiver/kafkametricsreceiver/README.md
Comment (crobert-1) [inline receiver/kafkametricsreceiver/README.md:38]: To make sure I understand: The goal of this configuration option is to be able to set an attribute value for all relevant metrics that are emitted by this receiver, is that correct? 

If so, is there a reason we're not relying on a processor to add the attribute? Also, should this actually a resource attribute, rather then datapoint attribute?

**Comment by crobert-1** [comment]
Hello @RutvikS-crest, please see CI/CD failures as well as it appears they're related to this change.

**Comment by RutvikS-crest** [comment]
> Hello @RutvikS-crest, please see CI/CD failures as well as it appears they're related to this change.

Hi @crobert-1 Fixed the pipeline failures, except the CHANGELOG.

File: receiver/kafkametricsreceiver/README.md
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/README.md:38]: Correct, It's intended to provide a new attribute (`cluster_alias`) for all metrics. It will store a cluster identifier.
It's possible to achieve this using processors, But it's hard to manage when we have many clusters. for current implementation we need to maintain separate pipelines for each clusters. 

for example,
```yaml
...

processors:
  resource/kafka-1-label:
    attributes:
      - key: cluster_alias
        action: insert
        value: kafka-1
  resource/kafka-2-label:
    attributes:
      - key: cluster_alias
        action: insert
        value: kafka-2

service:
  pipelines:
    metrics/kafka-1:
      receivers:
        - kafkametrics/kafka-1
      processors: [resource/kafka-1-label]
      exporters: [signalfx]
    metrics/kafka-2:
      receivers:
        - kafkametrics/kafka-2
      processors: [resource/kafka-2-label]
      exporters: [signalfx]
```

it would be great if there is an optional cluster identifier that makes filtering metrics easier.

Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/README.md:38]: And yes, I have added resource attribute for `cluster_alias`.

**Comment by crobert-1** [comment]
From failing CI/CD action:
```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```

**Comment by RutvikS-crest** [comment]
> From failing CI/CD action:
> 
> ```
> Generated code is out of date, please run "make generate" and commit the changes in this PR.
> ```

Thanks, It's updated

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (crobert-1) [inline receiver/kafkametricsreceiver/metadata.yaml:43]: We generally prefer to not enable metrics by default. I see all of the added metrics are enabled by default, can you share justification as to why they need to be enabled by default? (For more context, it's considered a breaking change to add any new metrics that will be enabled by default).

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (crobert-1) [inline receiver/kafkametricsreceiver/metadata.yaml:45]: I believe from [semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-units) that the time unit of durations need to be in seconds.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (crobert-1) [inline receiver/kafkametricsreceiver/metadata.yaml:68]: I believe from [semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-units) this unit should be in seconds.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (crobert-1) [inline receiver/kafkametricsreceiver/metadata.yaml:67]: Since it's a kafka setting, is it valuable to include the default here? I'm worried this may go stale, or just not provide useful information to the user.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (crobert-1) [inline receiver/kafkametricsreceiver/metadata.yaml:42]: For all of the metrics being added: The unit shouldn't be included in the metric name. [Source](https://opentelemetry.io/docs/specs/semconv/general/metrics/#units)

**Review by crobert-1 — COMMENTED** [review]
I've only reviewed the `metadata.yaml` at this point as other parts of this PR will change as a result of discussion there. 👍

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Thanks for the context, It's fine if we keep these new metrics disabled by default. I have made necessary changes.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Thanks, I have changed the units to follow standard norms, Also made appropriate time conversions from logic side.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Thanks, I have changed the units to follow standard norms, Also made appropriate time conversions from logic side.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Sure, Removed that default value from the docs, I am anyways not adding any logic to set these defaults, Kafka takes care of that.

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Thanks for these docs, I have made necessary changes. Please LMK if any of these looks odd.

File: receiver/kafkametricsreceiver/README.md
Comment (crobert-1) [reply receiver/kafkametricsreceiver/README.md:38]: Okay, I could see the potential value there. I'll defer to code owners for a more definitive response, but it's not a blocker for me.

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/broker_scraper.go:68]: Question out of ignorance, since `clusterAdmin` is a value variable, not a pointer, can it ever be nil?

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/broker_scraper.go:71]:
```suggestion
			s.settings.Logger.Error("Error creating kafka client with admin privileges", zap.Error(err))
```

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/broker_scraper.go:82]: I'm a bit confused, is this string a kafka-specific value? Unrelated to the metrics in this receiver? The value is being used to create a metric with a different name, so I just want to make sure I'm following this properly.

File: receiver/kafkametricsreceiver/generated_package_test.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/generated_package_test.go:1]: I believe this change is incorrect, it can be fixed by running `make gofmt`

File: receiver/kafkametricsreceiver/scraper_test_helper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/scraper_test_helper.go:168]:
```suggestion
		return nil, fmt.Errorf("no config entries found for topic")
```

File: receiver/kafkametricsreceiver/scraper_test_helper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/scraper_test_helper.go:191]: Is this name purposefully different than what the scraper is looking for, `log.retention.hours`?

File: receiver/kafkametricsreceiver/scraper_test_helper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/scraper_test_helper.go:192]: The unit in the variable name here doesn't match the name. Is there a reason for the mismatch?

File: receiver/kafkametricsreceiver/topic_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/topic_scraper.go:120]: Out of ignorance, is there a reason this isn't being added to `errors`?

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/broker_scraper.go:68]: Yes It can, clusterAdmin is sarama.ClusterAdmin interface, It will be nil if it's not initialised.

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/broker_scraper.go:82]: Yes this string (and similar ones for topic) is a inbuilt config of a Kafka broker, We are just fetching this configs from kafka. I have avoided adding "hours" as metric name, Just to make it compliant with otel simentics. The unit for all time metrics is "s".

File: receiver/kafkametricsreceiver/scraper_test_helper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/scraper_test_helper.go:192]: My bad, Missed to update it in tests

File: receiver/kafkametricsreceiver/topic_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/topic_scraper.go:120]: No Particular reason, Just wanted to keep behaviour consistent with existing receiver.

**Comment by crobert-1** [comment]
@RutvikS-crest Can you fix the conflicts when you get a chance?

**Comment by RutvikS-crest** [comment]
> @RutvikS-crest Can you fix the conflicts when you get a chance?
@crobert-1 conflicts are resolved for now.

File: .chloggen/receiver-kafkametrics-topic-configs.yaml
Comment (crobert-1) [inline .chloggen/receiver-kafkametrics-topic-configs.yaml:10]:
```suggestion
note: "Add option to configure cluster alias name and add new metrics for kafka topic configurations"
```

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/broker_scraper.go:85]:
```suggestion
			scrapeErrors.AddPartial(1, fmt.Errorf("failed to fetch the `log.retention.hours` metric from %s: %w", broker.Addr(), err))
```
Nit

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/broker_scraper.go:94]:
```suggestion
				scrapeErrors.AddPartial(1, fmt.Errorf("error converting `log.retention.hours` for %s: value was %s", broker.Addr(), config.Value))
```
Nit: We should make it clear that the conversion is what failed, and what value failed to be converted. Alternatively, we may want to actually include the error returned from `strconv.Atoi`

File: receiver/kafkametricsreceiver/scraper_test_helper.go
Comment (crobert-1) [inline receiver/kafkametricsreceiver/scraper_test_helper.go:201]: We have a few kafka metric strings that are used multiple times that could instead be central variables, such as `min.insync.replicas`, `retention.ms`, `retention.bytes`, and `log.retention.hours`.

**Comment by crobert-1** [comment]
Looks like there are some CI failures, here's one example:
```
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:186:2: strMinInsyncReplicas declared and not used
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:187:2: strLogRetentionMs declared and not used
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:188:2: strLogRetentionBytes declared and not used
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:201:4: undefined: TOPIC_MIN_INSYNC_REPLICA
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:202:4: undefined: TOPIC_RETENTION_PERIOD
Error: ../../receiver/kafkametricsreceiver/scraper_test_helper.go:[20](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/10212728612/job/28272184082?pr=34149#step:6:21)3:4: undefined: TOPIC_RETENTION_SIZE
```
Please check to see if there are any others as well. 👍

**Comment by RutvikS-crest** [comment]
> Please check to see if there are any others as well. 👍

My bad, Didn't push one commit 🤦🏽‍♂️, Tests should pass now

**Review by crobert-1 — APPROVED** [review]
Thanks for all of your work here, and for your patience with the back and forth 👍

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (atoulme) [inline receiver/kafkametricsreceiver/metadata.yaml:15]: should this be enabled by default?

Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/metadata.yaml]: Yes, We should disable it by default, I did the same, It will only be enabled when the new `cluster_alias` field is set in the config.

**Comment by dmitryax** [comment]
In future please break down such big PR. I'll be easier to review and merge

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/broker_scraper.go:76]: I have returned these to ensure backward compatibility. This makes sure metrics recorded until this point are exported even if clusterAdmin initiation fails.

File: receiver/kafkametricsreceiver/consumer_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/consumer_scraper.go:165]: The SetClusterAlias function itself takes care of checking if the attribute is enabled, We shouldn't duplicate the logic here. ref: [generated_resource.go#L25](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1f6892bc60f6a49674f204d19626af15efa65843/receiver/kafkametricsreceiver/internal/metadata/generated_resource.go#L25)

File: receiver/kafkametricsreceiver/topic_scraper.go
Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/topic_scraper.go:76]: added a logic to avoid unnecessary scrapping in the scrapeTopicConfigs function itself.

File: receiver/kafkametricsreceiver/consumer_scraper.go
Comment (dmitryax) [reply receiver/kafkametricsreceiver/consumer_scraper.go:165]: Ok, the condition is based on the enabled flag. It becomes a bit verbose to enable it in both config sections. I'm fine with that for now

File: receiver/kafkametricsreceiver/broker_scraper.go
Comment (dmitryax) [reply receiver/kafkametricsreceiver/broker_scraper.go:79]: ok, sounds good

File: receiver/kafkametricsreceiver/metadata.yaml
Comment (dmitryax) [inline receiver/kafkametricsreceiver/metadata.yaml:13]: Let's call it `kafka.cluster.alias` to be more OTel compliant

File: receiver/kafkametricsreceiver/README.md
Comment (dmitryax) [inline receiver/kafkametricsreceiver/README.md:38]: it's added as _resource_ attribute

Comment (RutvikS-crest) [reply receiver/kafkametricsreceiver/README.md]: revised the language let me know if this doesn't work.

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
Please run `make generate` to pass the CI

**Pull request merged**
